### ISSUE TITLE
torch.fft.rfft2 doesn't support half dtype

### ISSUE THREAD

**Issue by dong03** [pytorch/pytorch#70664]
### 🐛 Describe the bug

Hi, I try to use both **torch.fft.rfft2** and half-precision (via **torch.cuda.amp**), and they seem don't work together.

(For an easy demonstration, I directly assign half precision via dtype.
```python
import torch
a = torch.randn((2,3,128,128), dtype=torch.float16)
torch.fft.rfft2(a, norm='backward')
```

Output:

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-19-b4a82a9b573a> in <module>
      1 import torch
      2 a = torch.randn((2,3,128,128), dtype=torch.float16)
----> 3 torch.fft.rfft2(a, norm='backward')

RuntimeError: Unsupported dtype Half
```

### Versions

torch==1.8.1

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @peterbell10

**Comment by peterbell10** [comment]
This requires better `complex32` support.

**Comment by Pondiniii** [comment]
Same problem, any fix?

**Comment by ezyang** [comment]
see https://github.com/pytorch/pytorch/issues/71680

**Comment by louvinci** [comment]
This problem is solved in the version of pytorch>=1.12.0. 
See the official [instructions](https://pytorch.org/docs/1.12/generated/torch.fft.rfft2.html#torch.fft.rfft2)
```
Supports torch.half on CUDA with GPU Architecture SM53 or greater. However it only supports powers of 2 signal length in every transformed dimensions.
```
Here, the underlying GPU is A100
```
import torch
a = torch.randn((4,4), dtype=torch.float16).cuda()
b = a.float()
ffta = torch.fft.rfft2(a, norm='backward')
fftb = torch.fft.rfft2(b, norm='backward')
fftb-ffta
```
output
```
tensor([[-0.0013+0.0000j, -0.0012+0.0030j,  0.0027+0.0000j],
        [ 0.0003-0.0005j,  0.0002+0.0004j,  0.0012+0.0034j],
        [-0.0022+0.0000j,  0.0017-0.0009j, -0.0022+0.0000j],
        [ 0.0003+0.0005j, -0.0007-0.0005j,  0.0012-0.0034j]], device='cuda:0')

```
Can be executed rightly

**Comment by dong03** [comment]
Many thx for kind reminder :)

**Comment by Jup11ter** [comment]
I have observed a similar buggy behavior on `torch.fft.ifft` in PyTorch 1.10.0 version. The following codes will raise
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    torch.fft.ifft(a, norm='backward')
RuntimeError: Unsupported dtype Half
```
Luckily, this support problem of dtype is solved in 1.12.0 version.
I think the fix on `torch.fft.ifft` may also be related to this [commit](https://github.com/pytorch/pytorch/commit/9f7268f5b12e4a7c108efb1219a52fbc67709cd9).

```

import torch
a = torch.randn((2,3,128,128), dtype=torch.float16).cuda()
torch.fft.ifft(a, norm='backward')
```

**Comment by wuhaoupup** [comment]
x = x.to(dtype=torch.float32)
y = torch.fft.rfft2(x, norm=self.norm)

put the dtype to float32,
in my code, its right

**Comment by Billy-ZTB** [comment]
> This problem is solved in the version of pytorch>=1.12.0. See the official [instructions](https://pytorch.org/docs/1.12/generated/torch.fft.rfft2.html#torch.fft.rfft2)
> 
> ```
> Supports torch.half on CUDA with GPU Architecture SM53 or greater. However it only supports powers of 2 signal length in every transformed dimensions.
> ```
> 
> Here, the underlying GPU is A100
> 
> ```
> import torch
> a = torch.randn((4,4), dtype=torch.float16).cuda()
> b = a.float()
> ffta = torch.fft.rfft2(a, norm='backward')
> fftb = torch.fft.rfft2(b, norm='backward')
> fftb-ffta
> ```
> 
> output
> 
> ```
> tensor([[-0.0013+0.0000j, -0.0012+0.0030j,  0.0027+0.0000j],
>         [ 0.0003-0.0005j,  0.0002+0.0004j,  0.0012+0.0034j],
>         [-0.0022+0.0000j,  0.0017-0.0009j, -0.0022+0.0000j],
>         [ 0.0003+0.0005j, -0.0007-0.0005j,  0.0012-0.0034j]], device='cuda:0')
> ```
> 
> Can be executed rightly

But when using fftshift after rfftn, an error accur: RuntimeError: "roll_cuda" not implemented for 'ComplexHalf'
